core/include/gnuradio-4.0/HistoryBuffer.hpp

```diff
@@ -100,23 +113,176 @@ class HistoryBuffer {
     /**
      * @brief Adds a range of elements the end expiring the oldest elements beyond the buffer's capacities.
      */
-    template<typename Iter>
-    constexpr void push_back_bulk(Iter cbegin, Iter cend) noexcept {
+    template<std::input_iterator Iter>
+    constexpr void push_back(Iter cbegin, Iter cend) noexcept {
         const auto nSamplesToCopy = std::distance(cbegin, cend);
         Iter       optimizedBegin = static_cast<std::size_t>(nSamplesToCopy) > _capacity ? std::prev(cend, static_cast<std::ptrdiff_t>(_capacity)) : cbegin;
         for (auto it = optimizedBegin; it != cend; ++it) {
             push_back(*it);
         }
     }
 
+    template<std::input_iterator Iter>
+    constexpr void push_back_bulk(Iter cbegin, Iter cend) noexcept {
+        return push_back(cbegin, cend);
+    }
+
     /**
      * @brief Adds a range of elements the end expiring the oldest elements beyond the buffer's capacities.
      */
-    template<typename Range>
+    template<std::ranges::range Range>
+    constexpr void push_back(const Range& range) noexcept {
+        push_back_bulk(range.cbegin(), range.cend());
+    }
+
+    template<std::ranges::range Range>
     constexpr void push_back_bulk(const Range& range) noexcept {
         push_back_bulk(range.cbegin(), range.cend());
     }
 
+    /**
+     * @brief Adds an element to the front expiring the oldest element beyond the buffer's capacities.
```

Comment:
Not sure what this means.

Reply:
It's still a circular/limited buffer size. If you add a sample (either front or back) the oldest is being dropped once the maximum size is reached. This is the difference to the `std::deque` behaviour.

Reply:
No, I get that, just wanted to say the comment itself is not that clear :)